-1

Tested at: a540a1a208902b708f8c42e5be52befe45114089

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-17525/17800/summary.html

I found follow errors while testing this PR

**Failed tests:  Build  ClangBuild**

* **Build**:

I found an error when building: 
 
 <pre>In file included from /cvmfs/cms-ib.cern.ch/nweek-02459/slc6_amd64_gcc530/external/boost/1.57.0-oenich2/include/boost/lexical_cast/detail/converter_lexical_streams.hpp:64:
In file included from /cvmfs/cms-ib.cern.ch/nweek-02459/slc6_amd64_gcc530/external/boost/1.57.0-oenich2/include/boost/lexical_cast/detail/inf_nan.hpp:34:
In file included from /cvmfs/cms-ib.cern.ch/nweek-02459/slc6_amd64_gcc530/external/boost/1.57.0-oenich2/include/boost/math/special_functions/sign.hpp:16:
In file included from /cvmfs/cms-ib.cern.ch/nweek-02459/slc6_amd64_gcc530/external/boost/1.57.0-oenich2/include/boost/math/tools/config.hpp:335:
In file included from /cvmfs/cms-ib.cern.ch/nweek-02459/slc6_amd64_gcc530/external/boost/1.57.0-oenich2/include/boost/detail/fenv.hpp:97:
/cvmfs/cms-ib.cern.ch/nweek-02459/slc6_amd64_gcc530/external/gcc/5.3.0/include/c++/5.3.0/fenv.h:58:11: error: no member named 'fenv_t' in the global namespace
  using ::fenv_t;
        ~~^
/cvmfs/cms-ib.cern.ch/nweek-02459/slc6_amd64_gcc530/external/gcc/5.3.0/include/c++/5.3.0/fenv.h:59:11: error: no member named 'fexcept_t' in the global namespace
  using ::fexcept_t;
        ~~^
</pre>

* **Clang**:

I found a compilation error while trying to compile with clang: 
I used this command: 
 scram b vclean && scram build -k -j 48 USER_CXXFLAGS='-fsyntax-only' COMPILER='llvm compile'


 <pre>>> Compiling edm plugin /build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_0_X_2017-02-15-1100/src/Geometry/HcalEventSetup/src/CaloTowerHardcodeGeometryEP.cc 
>> Compiling edm plugin /build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_0_X_2017-02-15-1100/src/Geometry/HcalEventSetup/src/HcalAlignmentEP.cc 
>> Compiling edm plugin /build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_0_X_2017-02-15-1100/src/Geometry/HcalEventSetup/src/SealModule.cc 
>> Compile sequence completed for CMSSW CMSSW_9_0_X_2017-02-15-1100
+ echo ALL_OK
ALL_OK
</pre>

--- a/Geometry/HcalCommonData/interface/HcalDDDRecConstants.h
+++ b/Geometry/HcalCommonData/interface/HcalDDDRecConstants.h
@@ -114,7 +114,7 @@ public:
 private:
   void                      getOneEtaBin(int subdet, int ieta, int zside,
 					 std::vector<std::pair<int,double>>& phis,
-					 std::map<int,int>& layers, bool flag,
+					 std::map<int,int>& layers, bool planOne,
 					 std::vector<HcalDDDRecConstants::HcalEtaBin>& bins) const;
   void                      initialize(void);
   unsigned int              layerGroupSize(int eta) const;

--- a/Geometry/HcalCommonData/interface/HcalDDDSimConstants.h
+++ b/Geometry/HcalCommonData/interface/HcalDDDSimConstants.h
@@ -40,7 +40,7 @@ public:
 					  const int zside) const;
   int                       getDepthEta16M(const int det) const;
   int                       getDepthEta29(const int phi, int zside, int i) const;
-  int                       getDepthEta29M(const int i, const bool flag) const;
+  int                       getDepthEta29M(const int i, const bool planOne) const;
   std::pair<int,double>     getDetEta(const double eta, const int depth);
   int                       getEta(const int det, const int lay, const double hetaR);
   std::pair<int,int>        getEtaDepth(const int det, int etaR, int phi,

--- a/Geometry/HcalCommonData/src/HcalDDDRecConstants.cc
+++ b/Geometry/HcalCommonData/src/HcalDDDRecConstants.cc
@@ -579,7 +579,7 @@ void HcalDDDRecConstants::specialRBXHBHE(const std::vector<HcalDetId>& idsOld,
 
 void HcalDDDRecConstants::getOneEtaBin(int subdet, int ieta, int zside,
 				       std::vector<std::pair<int,double> >& phis,
-				       std::map<int,int>& layers, bool flag,
+				       std::map<int,int>& layers, bool planOne,
 				       std::vector<HcalDDDRecConstants::HcalEtaBin>& bins) const {
 
   unsigned int lymax = (subdet == 1) ? 17 : 19;
@@ -630,7 +630,7 @@ void HcalDDDRecConstants::getOneEtaBin(int subdet, int ieta, int zside,
 	if (subdet == 1 && ieta == iEtaMax[subdet-1] 
 	    && dep > hcons.getDepthEta16M(1)) break;
 	if (subdet == 2 && ieta == hpar->noff[1] &&  
-	    dep > hcons.getDepthEta29M(0,flag)) {
+	    dep > hcons.getDepthEta29M(0,planOne)) {
 	  lmax = lymx0;
 	  break;
 	}

--- a/Geometry/HcalCommonData/src/HcalDDDSimConstants.cc
+++ b/Geometry/HcalCommonData/src/HcalDDDSimConstants.cc
@@ -188,9 +188,10 @@ int HcalDDDSimConstants::getDepthEta29(const int phi, const int zside,
   return depth;
 }
 
-int HcalDDDSimConstants::getDepthEta29M(const int i, const bool flag) const {
+int HcalDDDSimConstants::getDepthEta29M(const int i, 
+					const bool planOne) const {
   int depth = (i == 1) ? depthEta29[1] : depthEta29[0];
-  if (i == 0 && flag) {
+  if (i == 0 && planOne) {
     std::vector<int> phis;
     int detsp = ldmap_.validDet(phis);
     if (detsp == 2) {

--- a/Geometry/HcalTowerAlgo/test/HcalDetIdTester.cc
+++ b/Geometry/HcalTowerAlgo/test/HcalDetIdTester.cc
@@ -1,0 +1,95 @@
+#include "FWCore/Framework/interface/one/EDAnalyzer.h"
+#include "FWCore/Framework/interface/EventSetup.h"
+#include "FWCore/Framework/interface/ESHandle.h"
+#include "FWCore/Framework/interface/MakerMacros.h"
+#include "Geometry/CaloGeometry/interface/CaloGeometry.h"
+#include "Geometry/CaloGeometry/interface/CaloCellGeometry.h"
+#include "Geometry/HcalCommonData/interface/HcalDDDRecConstants.h"
+#include "Geometry/HcalTowerAlgo/interface/HcalGeometry.h"
+#include "Geometry/HcalTowerAlgo/interface/HcalFlexiHardcodeGeometryLoader.h"
+#include "Geometry/Records/interface/CaloGeometryRecord.h"
+#include <iostream>
+
+class HcalDetIdTester : public edm::one::EDAnalyzer<> {
+
+public:
+  explicit HcalDetIdTester( const edm::ParameterSet& );
+  ~HcalDetIdTester( void );
+    
+  void beginJob() override {}
+  void analyze(edm::Event const& iEvent, edm::EventSetup const&) override;
+  void endJob() override {}
+
+private:
+  edm::ParameterSet ps0_;
+  bool              geomDB_;
+};
+
+HcalDetIdTester::HcalDetIdTester( const edm::ParameterSet& iConfig ) : ps0_(iConfig) {
+  geomDB_ = iConfig.getParameter<bool>("GeometryFromDB");
+}
+
+HcalDetIdTester::~HcalDetIdTester( void ) {}
+
+void
+HcalDetIdTester::analyze(const edm::Event& /*iEvent*/, 
+			 const edm::EventSetup& iSetup ) {
+
+  edm::ESHandle<HcalDDDRecConstants> hDRCons;
+  iSetup.get<HcalRecNumberingRecord>().get(hDRCons);
+  const HcalDDDRecConstants hcons = (*hDRCons);
+  edm::ESHandle<HcalTopology> topologyHandle;
+  iSetup.get<HcalRecNumberingRecord>().get( topologyHandle );
+  const HcalTopology topology = (*topologyHandle);
+
+  CaloSubdetectorGeometry* caloGeom(0);
+  if (geomDB_) {
+    edm::ESHandle<CaloGeometry> pG;
+    iSetup.get<CaloGeometryRecord>().get(pG);
+    const CaloGeometry* geo = pG.product();
+    caloGeom = (CaloSubdetectorGeometry*)(geo->getSubdetectorGeometry(DetId::Hcal,HcalBarrel));
+  } else {
+    HcalFlexiHardcodeGeometryLoader m_loader(ps0_);
+    caloGeom = m_loader.load(topology, hcons);
+  }
+
+  int maxDepth = (topology.maxDepthHB() > topology.maxDepthHE()) ?
+    topology.maxDepthHB() : topology.maxDepthHE();
+
+  int nfail0(0);
+  for (int det = 1; det <= HcalForward; det++) {
+    for (int eta = -HcalDetId::kHcalEtaMask2; 
+	 eta <= HcalDetId::kHcalEtaMask2; eta++) {
+      for (int depth = 1; depth <= maxDepth; depth++) {
+	for (int phi = 0; phi <= HcalDetId::kHcalPhiMask2; phi++) {
+	  HcalDetId detId ((HcalSubdetector) det, eta, phi, depth);
+	  if (topology.valid(detId)) {
+	    const CaloCellGeometry * cell = caloGeom->getGeometry((DetId)(detId));
+	    if (cell) {
+	      std::cout << detId << " " << cell->getPosition() << std::endl;
+	    } else {
+	      std::cout << detId << " position not found" << std::endl;
+	      ++nfail0;
+	    }
+	  }
+	}
+      }
+    }
+  }
+
+  int nfail1(0);
+  const std::vector<DetId>& ids = caloGeom->getValidDetIds();
+  for (std::vector<DetId>::const_iterator itr = ids.begin();  itr != ids.end();
+       ++itr)  {
+    if (!topology.valid(*itr)) {
+      std::cout << HcalDetId(*itr) << " declared as invalid == ERROR\n";
+      ++nfail1;
+    }
+  }
+
+  std::cout << "\nNumber of failures:\nTopology certifies but geometry fails " 
+	    << nfail0 << "\nGeometry certifies but Topology fails " << nfail1
+	    << std::endl << std::endl;
+}
+
+DEFINE_FWK_MODULE(HcalDetIdTester);

--- a/Geometry/HcalTowerAlgo/test/runHcalDetIdTester_cfg.py
+++ b/Geometry/HcalTowerAlgo/test/runHcalDetIdTester_cfg.py
@@ -1,0 +1,19 @@
+import FWCore.ParameterSet.Config as cms
+
+process = cms.Process("HcalGeometryTest")
+
+process.load("Geometry.HcalCommonData.testGeometry17bXML_cfi")
+process.load("Geometry.HcalCommonData.hcalDDConstants_cff")
+process.load("Geometry.HcalEventSetup.hcalTopologyIdeal_cfi")
+
+process.source = cms.Source("EmptySource")
+process.maxEvents = cms.untracked.PSet(
+    input = cms.untracked.int32(1)
+    )
+
+process.hga = cms.EDAnalyzer("HcalDetIdTester",
+                             GeometryFromDB = cms.bool(False))
+
+process.Timing = cms.Service("Timing")
+
+process.p1 = cms.Path(process.hga)

